### PR TITLE
fixing compile-time warnings

### PR DESCRIPTION
**Pull request by ipdemes**

### CODE REVIEW

**Comment by magnatelee** [comment]
@ipdemes make sure you address all issues @manopapad raised in #341.

**Comment by ipdemes** [comment]
@magnatelee : I had some issues with my workstation and submitted this PR before finishing addressing the comments. I should have added WIP. 
@manopapad : I believe I have addressed all issues mentioned in #341 here

Note that my original approach of using partial template specialization(first commit here) didn't work. It seems like some `CuNumericCodeTypes` objects gets converted to `LegateCodeType` objects when stored/asked for in the type storage (on the Legate side). So I implemented some workaround converting `code` to `size_t` for `CunumericCodeTypes`.

File: cunumeric/deferred.py [hunk 1]
```diff
@@ -633,30 +633,19 @@ def get_item(self, key):
...
             if copy_needed:
                 # Create a new array to be the result
-                result = self.runtime.create_empty_thunk(
-                    index_array.base.shape,
+                result_store = self.context.create_store(
                     self.dtype,
-                    inputs=[self],
+                    shape=index_array.shape,
+                    optimize_scalar=False,
+                )
```
Comment (ipdemes) [inline cunumeric/deferred.py:640]: @magnatelee Is it OK to keep `optimize_scalar=False` for both future and non-future cases?

Comment (magnatelee) [reply cunumeric/deferred.py:640]: we want to use scalars wherever applicable, so I don't want us to pass `optimize_scalar=False` blindly

**Comment by magnatelee** [comment]
> Note that my original approach of using partial template specialization(first commit here) didn't work. It seems like some CuNumericCodeTypes objects gets converted to LegateCodeType objects when stored/asked for in the type storage (on the Legate side). So I implemented some workaround converting code to size_t for CunumericCodeTypes.

I was actually surprised to see that you were able to chain the two dispatch methods despite their using different code types. I feel we shouldn't even try the chaining and should rather have an if statement outside and keep the cunumeric dispatch only working on CuNumericCodeType. Going through size_t to make it work seems like a hack.

**Comment by magnatelee** [comment]
> I have tried that, but it didn't work. It seems like somewhere in the Legate (I assume when we register new type) the type of the cunumeric types changes from CuNumericCodeType to LegateCodeType. I will investigate it further and will try to find more details on this.

I think the culprit is this line: https://github.com/nv-legate/legate.core/blob/branch-22.05/src/core/data/store.h#L284
I thought I changed the code to be an arbitrary 32-bit integer, but looks like I haven't done it yet. I'll push a fix.

**Comment by ipdemes** [comment]
@magnatelee : sounds good!

**Comment by magnatelee** [comment]
@ipdemes https://github.com/nv-legate/legate.core/pull/239 note that you will still need to hoist the if condition on the type codes with this change, as `CunumericCodeTypes` and `LegateTypeCode` are different types and have disjoint type codes.

**Review by magnatelee — APPROVED** [review]

**Pull request merged**